### ISSUE TITLE
ReadNextAsync throwing  'Error reading object reference '1' serialization error when using V3 SDK and Newtonsoft JSON with settings PreserveReferencesHandling.All

### ISSUE THREAD

**Issue by CraZZyEddie** [Azure/azure-cosmos-dotnet-v3#3077]
We are getting an exception 'Error reading object reference '1' from Newtonsoft caused by having documents serialized with JsonSerializerSettings PreserveReferencesHandling enabled.

It seems that the new Cosmos V3 simply places the resulting JSON from each Item into a new document setup as an array.
This creates invalid Newtonsoft Json because each document contains references ($id) that are local to it.
When all the JSON is combined nothing fixes the document refences when everything is placed into a single document.

The error from Newtonsoft then makes sense "Error reading object reference '1'. Path".
Its complaining about a duplicate reference in the document.
Newtonsoft stores the references as $id which you will notices many repeats of this in the data from Cosmos.

I am using the JSON serializer from Microsoft on Github.
[https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Serializer/CosmosJsonDotNetSerializer.cs](url)

You can see all this by adding some code to CosmosJsonDotNetSerializer . FromStream.
Before the JSON is de-serialized use var myTemp = sr.ReadToEnd() to see the stream data.
Remember to reset it after this.. stream.Position=0 if you want things to keep going.

After more experimentation I did find that MetadataPropertyHandling being turned off fixes the Exception. However, the MetadataPropertyHandling has more than identifiers in the Metadata, it also contains the stored type in the $type property.

We use this interesting technique in C# called inheritance.
When data is read in at times there are objects of multiple types retuned in a query.
With the MetadataPropertyHandline disabled every object created is of the type requested and not the type that was encoded into the metadata.

If your query was for Shapes and returns a Square, Circle and Triangle then you will only get Shape type object back. If Shape is marked abstract the de-serialize will fail with another Exception.

I can now replace the two different API's with a few key strokes.
This code worked in the old API.

Our code needs to be able to request base types and get the actual type that was stored.

In summary, I would imagine the old API returns an array of strings with multiple blocks of JSON which are de-serialized independently.  It looks like someone felt they could get some speed back by joining the JSON and passing a single block back for de-serialization.   It easy to see the performance optimization in de-serializing as a single block.  However, as you support custom serializers it seems you are making too many assumptions that the JSON can be combined without affecting the de-serialization.

**I would suggest at the very least you should de-serialize differently if a Custom serializer is used.**

Thanks for any help you may offer.

Chris

**Comment by j82w** [comment]
Hi Chris,

The benchmark results showed by using an array instead of parsing and passing individual items reduced memory up to %50 and CPU utilization up to 25%. That is significant gains which justified the change. In most scenarios a JSON serializer automatically handle parsing an array of objects. 

Is there any chance you can provide a simple console app to better understand the scenario? The SDK and service return valid JSON so Netwonsoft should be able to handle it correctly.  This still seems like newtonsoft issue rather than a problem with the Cosmos DB SDK.

**Comment by CraZZyEddie** [comment]
After looking at code serialized with System.Text.Json which also allows reference handling.  It would seem like this would be an issue with any JSON serializer that adds reference tags into the JSON including System.Text.Json.

**Comment by CraZZyEddie** [comment]
I know this is a stretch but would it be possible to look at the JSON and if reference tags are seen adjust this optimization?
Even if examining the JSON is too expensive, you could provide a setting, to choose a path that avoids this optimization and allows code using refence handling?  Currently this leaves our system broke and in needs of architectural changes.

**Comment by iainx** [comment]
Here is a C# program that can reproduce this effect.

```c#
class Program
{
    public class TestObject
    {
        public string Value { get; set; }
        public string id { get; set; }
        public string partitionKey => "key";
    }
    
    static async Task Main(string[] args)
    {
        // Create a CosmosClient with a custom Newtonsoft JSON serializer
        var dbClient = new CosmosClient("AccountEndpoint=https://localhost:8081/;AccountKey=<snip>==",
            new CosmosClientOptions
            {
                Serializer = new CosmosJsonDotNetSerializer(new JsonSerializerSettings
                {
                    TypeNameHandling = TypeNameHandling.All,
                    ReferenceLoopHandling = ReferenceLoopHandling.Serialize,
                    PreserveReferencesHandling = PreserveReferencesHandling.All,
                    MetadataPropertyHandling = MetadataPropertyHandling.Default,
                    ConstructorHandling = ConstructorHandling.AllowNonPublicDefaultConstructor,
                    Converters = new JsonConverter[]
                    {
                        new StringEnumConverter()
                    }
                })
            });

        var database = dbClient.GetDatabase("issue3077");
        var container = database.GetContainer("items");

        var item = new TestObject { Value = "one", id = "one" };
        var response = await container.CreateItemAsync<TestObject>(item, new PartitionKey(item.partitionKey));

        item = new TestObject { Value = "two", id = "two" };
        response = await container.CreateItemAsync<TestObject>(item, new PartitionKey(item.partitionKey));

        var sql = "SELECT * FROM c OFFSET 0 LIMIT 50";
        var query = new QueryDefinition(sql);

        try 
        {
            var resultSet = container.GetItemQueryIterator<TestObject>(query);
            while (resultSet.HasMoreResults) 
            {
                var items = await resultSet.ReadNextAsync();
                foreach (var returnedItem in items) 
                {
                    Console.WriteLine($"Item: {returnedItem.Value}");
                }
            }
        }
        catch (Exception e) 
        {
            Console.WriteLine($"Exception {e}");
        }
    }
}

//------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
//------------------------------------------------------------

/// <summary>
/// The default Cosmos JSON.NET serializer.
/// </summary>
internal sealed class CosmosJsonDotNetSerializer : CosmosSerializer
{
       <snip>
}
```

the line `PreserveReferencesHandling = PreserveReferencesHandling.All` makes the `$id` metadata property be added to the resultant JSON document before it is sent to the database but because every item is serialized separately the `$id` value always starts at 1

This is the data for each item stored in CosmosDB

```json
{
    "$id": "1",
    "$type": "test_json.Program+TestObject, test-json",
    "Value": "one",
    "id": "one",
    "partitionKey": "key",
    "_rid": "-A4CANi9bC0PAAAAAAAAAA==",
    "_self": "dbs/-A4CAA==/colls/-A4CANi9bC0=/docs/-A4CANi9bC0PAAAAAAAAAA==/",
    "_etag": "\"00000000-0000-0000-88a3-11dcad1201d9\"",
    "_attachments": "attachments/",
    "_ts": 1684316194
}

{
    "$id": "1",
    "$type": "test_json.Program+TestObject, test-json",
    "Value": "two",
    "id": "two",
    "partitionKey": "key",
    "_rid": "-A4CANi9bC0QAAAAAAAAAA==",
    "_self": "dbs/-A4CAA==/colls/-A4CANi9bC0=/docs/-A4CANi9bC0QAAAAAAAAAA==/",
    "_etag": "\"00000000-0000-0000-88a3-11e3bbe901d9\"",
    "_attachments": "attachments/",
    "_ts": 1684316194
}
```

Ways around it would be either to disable `PreserveReferencesHandling` in the serializer which would leave the `$type` metadata, or to use the CosmosDB stream API to read the JSON and process it yourself to fix up or ignore the `$id`

IMO this is a case of bad data being put into the database, and the CosmosDB SDK can't be expected to fix up bad data on the way back out.
